# Parallelism in Computer Vision

## About this project

This project explores **Python parallelism** through a concrete computer vision use case: preprocessing and extracting features from images of the [Caltech-101](https://data.caltech.edu/records/mzrjq-6wc02) dataset (~9 000 images, 101 categories).

Each section introduces a parallelism concept, applies it to a real image-processing task, and benchmarks it against a serial baseline, so you can see exactly *when* and *why* parallelism helps.

---

### What you will learn

| Concept | Section |
|---|---|
| Measuring execution time reliably (mean over many runs) | [1.4 - Tools](#1.4---tools) |
| Serial preprocessing as a baseline | [2.2 - Serial](#2.2---serial) |
| `ProcessPoolExecutor`: true parallelism, bypassing the GIL | [2.3 - ProcessPoolExecutor](#2.3---processpoolexecutor) |
| `ThreadPoolExecutor`: concurrency limited by the GIL | [2.4 - ThreadPoolExecutor](#2.4---threadpoolexecutor) |
| Why image size affects preprocessing speed | [2.2 - Performance test](#2.2---serial) |
| Overhead vs throughput: why batching matters | [2.6 - Batch Preprocessing](#2.6---batch-preprocessing-using-processpoolexecutor) |
| Parallelising CNN feature extraction | [3 - Feature extraction](#3---feature-extraction) |
| Producer–consumer pipeline with `multiprocessing.Queue` | [5 - End-to-End Pipeline](#5---end-to-end-pipeline) |

---

### Notebook structure

```
1. Setup               : imports, environment detection, dataset, utility functions
2. Preprocessing       : serial → processes → threads → batched processes → speedup comparison
3. Feature extraction  : MobileNetV2 features via parallel batch processing
4. Visualisation       : PCA projection & KMeans clustering of CNN features
5. End-to-End Pipeline : CPU loader processes feeding a single model inference loop
```

---

### Key concepts in a nutshell

**GIL (Global Interpreter Lock)**: Python only lets *one thread* execute bytecode at a time.  
→ For CPU-bound work, threads do not speed things up, you need *processes*.  
→ For I/O-bound work (disk, network), threads are effective because threads can release the GIL while waiting.

**`ProcessPoolExecutor`**: spawns independent worker processes, each with its own Python interpreter and memory space. Bypasses the GIL entirely. Best for CPU-bound tasks.

**`ThreadPoolExecutor`**: lightweight worker threads sharing the same process memory. Affected by the GIL for CPU-bound code, but useful for I/O-bound tasks (e.g. downloading files, reading from disk).

**Batching**: sending a *chunk* of items to each worker instead of one item at a time. Reduces the number of inter-process messages, lowers scheduling overhead, and improves CPU cache locality.

**Producer–consumer pipeline**: dedicated processes load and decode images (*producers*) and push results onto a shared queue; a single process runs the ML model (*consumer*). Hides I/O latency behind compute.

## 1 - Setup

### 1.1 - Imports

In [1]:
import os
import time
import matplotlib.pyplot as plt
import numpy as np
import tensorflow as tf

# Image loading
from PIL import Image
from pathlib import Path
import zipfile
import urllib.request
from typing import List

# Parallel processing
from concurrent.futures import ProcessPoolExecutor, ThreadPoolExecutor, as_completed
#from python_parallelism.processes import parallel_map_batched
# Image preprocessing
#from python_parallelism.image_preprocessing import preprocess_one, preprocess_one_min, preprocess_all_serial

2026-03-10 17:17:26.929761: I external/local_xla/xla/tsl/cuda/cudart_stub.cc:31] Could not find cuda drivers on your machine, GPU will not be used.
2026-03-10 17:17:28.650930: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
2026-03-10 17:17:33.163444: I external/local_xla/xla/tsl/cuda/cudart_stub.cc:31] Could not find cuda drivers on your machine, GPU will not be used.


### 1.2 - Google Colab

In [2]:
def is_running_on_colab() -> bool:
    """
    Checks if the code is running on Google Colab.
    
    Returns:
        bool: True if running on Google Colab, False otherwise.
    """
    try:
        if 'google.colab' in str(get_ipython()): # type: ignore
            return True
        else:
            return False
    except NameError:
        return False

# If running on Google Colab, set up the environment and the root path to use Google Drive for data storage and processing
if is_running_on_colab():
    print("Running on Google Colab")

    # Check for available GPUs and, if found, configure TensorFlow to use memory growth
    physical_devices = tf.config.list_physical_devices('GPU')
    try:
        # Ensure at least one GPU is available
        if physical_devices:
            tf.config.experimental.set_memory_growth(physical_devices[0], True)
            print("GPU memory growth enabled.")
        else:
            print("No GPU devices found.")
    except:
        print(f"Error setting GPU memory growth: {e}") # type: ignore
        pass

    # Get GPU infos
    gpu_info = !nvidia-smi # type: ignore
    gpu_info = '\n'.join(gpu_info)
    if gpu_info.find('failed') >= 0:
        print('Not connected to a GPU')
    else:
        print(gpu_info)

    # Connect to Google Drive
    from google.colab import drive # type: ignore
    drive.mount('/content/drive')

    # Setup the Google root
    root_path = "./drive/MyDrive/Project/Python_parallelism"

    # Install jiwer for the calculate_average_wer function
    ! pip install jiwer
else:
    print("Running on a local machine")
    # Setup the local root
    root_path = ".."

# Displays the number of avaible GPUs
print("\nNum GPUs Available: ", len(tf.config.list_physical_devices('GPU')))


Running on a local machine

Num GPUs Available:  0


2026-03-10 17:17:36.763327: E external/local_xla/xla/stream_executor/cuda/cuda_platform.cc:51] failed call to cuInit: INTERNAL: CUDA error: Failed call to cuInit: UNKNOWN ERROR (303)


### 1.3 - Global Variables and Data paths for Caltech101 dataset

In [3]:
SAMPLE_SIZE = 1250

DATA_ROOT = Path(root_path, "data")
ARCHIVE_URL = "https://www.kaggle.com/api/v1/datasets/download/imbikramsaha/caltech-101"
ARCHIVE_PATH = DATA_ROOT / "caltech101.zip"
EXTRACT_DIR = DATA_ROOT / "caltech-101"

print(f"Data folder set to {DATA_ROOT}")


Data folder set to ../data


### 1.4 - Tools

This decorator can be used on any function if measuring the execution time is needed.

In [4]:
def time_it(func):
    """Decorator to measure execution time of functions"""
    def wrapper(*args, **kwargs):
        time_start = time.time()
        result = func(*args, **kwargs)
        time_end = time.time()
        print(f"{func.__name__} took {time_end - time_start:.4f}s")
        return result
    return wrapper

The following function will be used to measure the execution time of the preprocessing functions to come in this notebook, so we can compare them in the Evaluating Speedup section.

In [ ]:
def benchmark_function(func, *args, n_runs=5, **kwargs):
    """
    Run a function multiple times and return timing statistics.
    
    Args:
        func: Function to benchmark
        *args: Positional arguments to pass to func
        n_runs: Number of times to run the function (default: 5)
        **kwargs: Keyword arguments to pass to func
    
    Returns:
        dict: Contains 'mean', 'std', 'min', 'max', and 'n_runs' of the execution times
        list: List of execution times for each run
    """
    times = []
    for _ in range(n_runs):
        t0 = time.perf_counter()
        func(*args, **kwargs)
        t1 = time.perf_counter()
        times.append(t1 - t0)
    
    return {
        'mean': np.mean(times),
        'std': np.std(times),
        'min': np.min(times),
        'max': np.max(times),
        'n_runs': n_runs
    }, times

### 1.5 - Caltech101 dataset

In [6]:
def ensure_caltech101():
    if EXTRACT_DIR.exists() and any(EXTRACT_DIR.iterdir()):
        print("Caltech101 dataset already exists.")
        return
    print("Downloading Caltech101 archive...")
    urllib.request.urlretrieve(ARCHIVE_URL, ARCHIVE_PATH)
    print("Extracting Caltech101 archive...")
    with zipfile.ZipFile(ARCHIVE_PATH, "r") as zip_ref:
        zip_ref.extractall(path=DATA_ROOT)
    print("Caltech101 dataset extracted to", EXTRACT_DIR)

ensure_caltech101()


Caltech101 dataset already exists.


### 1.6 - Sample Images

In [7]:
def sample_images(root: Path, limit=None) -> List[Path]:
    """
    Create a list of image file paths from the given root directory.
    """
    extensions = {".jpg", ".jpeg", ".png", ".bmp"}
    paths = [path for path in root.rglob("*") if path.suffix.lower() in extensions]
    return paths[:limit] if limit else paths

paths = sample_images(EXTRACT_DIR, limit=None)
len(paths), paths[:5]


(9144,
 [PosixPath('../data/caltech-101/yin_yang/image_0042.jpg'),
  PosixPath('../data/caltech-101/yin_yang/image_0002.jpg'),
  PosixPath('../data/caltech-101/yin_yang/image_0053.jpg'),
  PosixPath('../data/caltech-101/yin_yang/image_0009.jpg'),
  PosixPath('../data/caltech-101/yin_yang/image_0019.jpg')])

## 2 - Preprocessing

### 2.1 - Main Pipeline: Resize, grayscale, Sobel edges.

Image resizing, grayscale conversion, and edges extraction with the [Sobel operator](https://homepages.inf.ed.ac.uk/rbf/HIPR2/sobel.htm).  
The grayscale conversion uses ITU-R BT.601 recommendation. Some explications [here](https://www.linkedin.com/pulse/why-grayscale-isnt-just-black-white-deep-dive-color-conversion-c-xidxf/). 

In [8]:
def load_resize(path, size=(224,224)):
    """
    Open an image, convert it to RGB to ensure we have 3 channels, and resize it.
    Returns a NumPy array of shape (height, width, 3).
    """
    with Image.open(path) as img:
        img = img.convert("RGB").resize(size, Image.BILINEAR)
        return np.array(img, dtype=np.uint8)

def to_gray(img_rgb):
    """
    Convert an RGB image to grayscale using perceptual weights.
    Returns a NumPy array of shape (height, width).
    """
    # Use of ellipsis (...) to select all pixels
    r, g, b = img_rgb[..., 0], img_rgb[..., 1], img_rgb[..., 2]
    # Apply ITU-R BT.601 to calculate luminance and cast float32 to avoid overflow later
    return (0.299 * r + 0.587 * g + 0.114 * b).astype(np.float32)

def sobel_edges(img_gray):
    """
    Apply the Sobel operator to a grayscale image of shape (height, width) to detect edges.
    Returns a NumPy array of the same shape as the input image.
    """
    # Create two 3x3 Sobel kernels
    sobel_x = np.array([[-1, 0, 1],
                        [-2, 0, 2],
                        [-1, 0, 1]], dtype=np.float32)
    sobel_y = np.array([[ 1, 2, 1],
                        [ 0, 0, 0],
                        [-1,-2,-1]], dtype=np.float32)

    # Pad the image to handle borders
    grad = np.pad(img_gray, ((1, 1), (1, 1)), mode='edge')
    # Initialize gradient images
    grad_x = (
        grad[:-2, :-2] * sobel_x[0, 0] + grad[:-2, 1:-1] * sobel_x[0, 1] + grad[:-2, 2:] * sobel_x[0, 2] +
        grad[1:-1, :-2] * sobel_x[1, 0] + grad[1:-1, 1:-1] * sobel_x[1, 1] + grad[1:-1, 2:] * sobel_x[1, 2] +
        grad[2:, :-2] * sobel_x[2, 0] + grad[2:, 1:-1] * sobel_x[2, 1] + grad[2:, 2:] * sobel_x[2, 2]
    )
    grad_y = (
        grad[:-2, :-2] * sobel_y[0, 0] + grad[:-2, 1:-1] * sobel_y[0, 1] + grad[:-2, 2:] * sobel_y[0, 2] +
        grad[1:-1, :-2] * sobel_y[1, 0] + grad[1:-1, 1:-1] * sobel_y[1, 1] + grad[1:-1, 2:] * sobel_y[1, 2] +
        grad[2:, :-2] * sobel_y[2, 0] + grad[2:, 1:-1] * sobel_y[2, 1] + grad[2:, 2:] * sobel_y[2, 2]
    )
    # Compute and return the gradient magnitude
    return np.sqrt(grad_x**2 + grad_y**2).astype(np.float32)

In [9]:
def preprocess_one(path, size=(224, 224), bins=32):
    """
    Preprocess pipeline for one image.
    Returns a dictionary containing components for reuse.
    """
    img_resized = load_resize(path, size)
    img_gray = to_gray(img_resized)
    img_edges = sobel_edges(img_gray)

    hists = []
    # Compute color histograms for each channel
    for channel in range(3):
        hist_channel, _ = np.histogram(img_resized[..., channel], bins=bins, range=(0, 255))
        hists.append(hist_channel.astype(np.float32))

    color_hist = np.concatenate(hists, axis=0)

    return {
        "path": str(path),
        "resized": img_resized,
        "gray": img_gray,
        "edges": img_edges,
        "color_hist": color_hist,
        "size": size,
        "bins": bins
    }

def preprocess_one_min(path, size=(224,224), bins=32):
    """
    Preprocess pipeline for one image.
    Returns a minimal dictionary of features.
    """
    img = load_resize(path, size)
    gray = to_gray(img)
    edges = sobel_edges(gray)
    # Simple feature: edge mean + color hist (density)
    hists = []
    # Compute color histograms for each channel
    for color in range(3):
        h, _ = np.histogram(img[..., color], bins=bins, range=(0,255), density=True)
        hists.append(h.astype(np.float32))
    color_hist = np.concatenate(hists)

    return {
        "path": str(path),
        "edge_mean": float(edges.mean()),
        "edge_max": float(edges.max()),
        "color_hist": color_hist
    }


### 2.2 - Serial  

Each image is preprocessed one at a time.


#### Full dictionary

Stores all intermediate arrays (the resized RGB image, the grayscale image, the Sobel edge map, and the colour histograms) plus the original path and processing parameters.  
Useful when you want to reuse or visualise intermediate results, but **memory-heavy**: each image contributes several large NumPy arrays.

In [ ]:
def preprocess_all_serial(paths, size=(224, 224), bins=32, limit=None):
    """
    Preprocess pipeline for a list of images.
    Returns a list of dictionaries, each containing components for reuse.
    """
    if limit is not None:
        paths = paths[:limit]
    return [preprocess_one(path, size=size, bins=bins) for path in paths]

bench_serial_full, times_serial_full = benchmark_function(preprocess_all_serial, paths, limit=SAMPLE_SIZE, n_runs=20)
bench_serial_full

# Uncomment to see the times of each run
# times_serial_full

#### Minimal dictionary

Stores only compact scalar features and the colour histogram, no full image arrays.  
Much lighter in memory and faster to pickle and send across process boundaries, making it a better fit for parallel workloads where workers need to return results to the main process.

In [ ]:
def preprocess_all_serial_min(paths, size=(224, 224), bins=32, limit=None):
    """
    Preprocess pipeline for a list of images.
    Returns a list of minimal dictionaries of features.
    """
    if limit is not None:
        paths = paths[:limit]
    return [preprocess_one_min(path, size=size, bins=bins) for path in paths]

bench_serial_min, times_serial_min = benchmark_function(preprocess_all_serial_min, paths, limit=SAMPLE_SIZE, n_runs=20)
bench_serial_min

# Uncomment to see the times of each run
# times_serial_min

#### Performance test on different image sizes

Image size directly affects both I/O time (bytes to decode) and compute time (pixels to process).  
Here we isolate these two components (`load_resize` vs `to_gray + sobel_edges`) across three resolutions to understand where time is actually spent before choosing a size for the parallel benchmarks.

In [16]:
import time, statistics as stats

def measure_components(paths, limit=60, size=(224,224)):
    time_load = []
    time_compute = []

    if not limit:
        limit = len(paths)

    for p in paths[:limit]:
        t0 = time.perf_counter()
        img = load_resize(p, size=size)
        t1 = time.perf_counter()
        gray = to_gray(img)
        sobel_edges(gray)
        t2 = time.perf_counter()
        time_load.append(t1 - t0)
        time_compute.append(t2 - t1)

    return {
        "n": limit,
        "load_mean": stats.mean(time_load),
        "compute_mean": stats.mean(time_compute),
        "load_total": sum(time_load),
        "compute_total": sum(time_compute)
    }


##### 224x224

In [ ]:
component_stats_224 = measure_components(paths, limit=SAMPLE_SIZE, size=(224,224))
component_stats_224

{'mean': np.float64(3.5650718285978655),
 'std': np.float64(0.656304665756646),
 'min': np.float64(3.339588986011222),
 'max': np.float64(6.381038499996066),
 'times': [6.381038499996066,
  3.8982742209918797,
  3.392423094017431,
  3.406361294008093,
  3.3722482929879334,
  3.3653467370022554,
  3.4553282729757484,
  3.3870120350038633,
  3.382667438010685,
  3.377185314981034,
  3.342259627999738,
  3.339588986011222,
  3.3799027430068236,
  3.4371985890029464,
  3.3521607129951008,
  3.4297769109834917,
  3.3540319220046513,
  3.367967038997449,
  3.42684010398807,
  3.4538247369928285],
 'n_runs': 20}

##### 112x112

In [ ]:
component_stats_112 = measure_components(paths, limit=SAMPLE_SIZE, size=(112,112))
component_stats_112

{'mean': np.float64(2.233498719848285),
 'std': np.float64(0.2761017781155716),
 'min': np.float64(2.020022039010655),
 'max': np.float64(3.167790926992893),
 'times': [2.3462623179948423,
  2.1581388149934355,
  2.0511020490084775,
  2.1074528629833367,
  2.0307582480018027,
  2.155192887992598,
  2.0489419999939855,
  2.0971027629857417,
  2.020022039010655,
  2.0893365669762716,
  2.153901672019856,
  2.073396685998887,
  2.0701457929972094,
  2.1300083860114682,
  2.0593451560125686,
  3.167790926992893,
  2.6758364340057597,
  2.4763847780122887,
  2.2769652199931443,
  2.4818887949804775],
 'n_runs': 20}

##### 56x56

In [ ]:
component_stats_56 = measure_components(paths, limit=SAMPLE_SIZE, size=(56,56))
component_stats_56

{'mean': np.float64(2.122290239400172),
 'std': np.float64(0.2582956724581338),
 'min': np.float64(1.7917569620185532),
 'max': np.float64(3.008777923998423),
 'times': [3.008777923998423,
  2.452222204999998,
  2.1016315770102665,
  1.9227481779817026,
  2.384491745993728,
  2.0284534659876954,
  2.1829039840085898,
  2.0817210789828096,
  1.9226917880005203,
  2.1193160169932526,
  1.9872784900071565,
  1.8997987120237667,
  2.2048041119996924,
  2.0664741480140947,
  2.074779179994948,
  1.9397233319759835,
  1.9072561959910672,
  1.7917569620185532,
  2.1750289800111204,
  2.1939467120100744],
 'n_runs': 20}

##### Speed Comparison

In [ ]:
for label, s in [("224x224", component_stats_224), ("112x112", component_stats_112), ("56x56", component_stats_56)]:
    total = s['load_total'] + s['compute_total']
    print(f"{label}:  load={s['load_total']:.3f}s  compute={s['compute_total']:.3f}s  total={total:.3f}s")

224x224:   3.5650718285978655
112x112:   2.233498719848285
56x56:     2.122290239400172


### 2.3 - ProcessPoolExecutor

Use of [ProcessPoolExecutor](https://docs.python.org/3/library/concurrent.futures.html#processpoolexecutor) from [concurrent.futures](https://docs.python.org/3/library/concurrent.futures.html#) to parallelise image preprocessing.

[ProcessPoolExecutor](https://docs.python.org/3/library/concurrent.futures.html#processpoolexecutor) spawns a pool of independent worker **processes**, each with its own Python interpreter and memory space.  
Because processes don't share the GIL, they can run CPU-bound code truly in parallel.  
Tasks are submitted to the pool and the executor distributes them among the workers.  
By default, the number of workers equals the number of CPU cores.

#### Full dictionary

In [ ]:
def preprocess_parallel(paths, size=(224,224), bins=32, max_workers=None, limit=None):
    """
    Preprocess pipeline using ProcessPoolExecutor for parallel processing.
    Returns a full dictionary of features.
    """
    if limit: paths = paths[:limit]
    out = []

    with ProcessPoolExecutor(max_workers=max_workers) as executor:
        futures = [executor.submit(preprocess_one, path, size, bins) for path in paths]
        for f in as_completed(futures):
            out.append(f.result())

    return out

bench_proc_full, times_proc_full = benchmark_function(preprocess_parallel, paths, limit=SAMPLE_SIZE, n_runs=3)
print(f"ProcessPool full dict:    mean={bench_proc_full['mean']:.4f}s  std={bench_proc_full['std']:.4f}s  (over {bench_proc_full['n_runs']} runs)")

# Uncomment to see the times of each run
# times_proc_full

#### Minimal dictionary

In [ ]:
def preprocess_parallel_min(paths, size=(224,224), bins=32, max_workers=None, limit=None):
    """
    Preprocess pipeline using ProcessPoolExecutor for parallel processing.
    Returns a minimal dictionary of features.
    """
    if limit: paths = paths[:limit]
    out = []

    with ProcessPoolExecutor(max_workers=max_workers) as executor:
        futures = [executor.submit(preprocess_one_min, path, size, bins) for path in paths]
        for f in as_completed(futures):
            out.append(f.result())

    return out

bench_proc_min, times_proc_min = benchmark_function(preprocess_parallel_min, paths, limit=SAMPLE_SIZE, n_runs=3)
print(f"ProcessPool minimal dict: mean={bench_proc_min['mean']:.4f}s  std={bench_proc_min['std']:.4f}s  (over {bench_proc_min['n_runs']} runs)")

# Uncomment to see the times of each run
# times_proc_min

### 2.4 - ThreadPoolExecutor

Use of [ThreadPoolExecutor](https://docs.python.org/3/library/concurrent.futures.html#threadpoolexecutor) from [concurrent.futures](https://docs.python.org/3/library/concurrent.futures.html#).

[ThreadPoolExecutor](https://docs.python.org/3/library/concurrent.futures.html#threadpoolexecutor) manages a pool of worker **threads** running inside the same process.  
Threads share memory (no data copying needed; they can access the same variables and objetcs), are cheap to create, and communicate easily.

However, because of the **GIL**, only one thread can execute Python bytecode at a time (concurrency).  
For CPU-bound work like image processing, threads therefore provide concurrency (interleaved execution) but **not parallelism** (simultaneous execution).  
We benchmark threads here specifically to observe this GIL effect and compare it against processes.

#### Full dictionary

In [ ]:
def preprocess_threads(paths, size=(224,224), bins=32, limit=None):
    """
    Preprocess pipeline using ThreadPoolExecutor.
    """
    if limit: paths = paths[:limit]

    with ThreadPoolExecutor() as executor:
        results = list(executor.map(lambda path: preprocess_one(path, size, bins), paths))

    return results

bench_threads_full, times_threads_full = benchmark_function(preprocess_threads, paths, limit=SAMPLE_SIZE, n_runs=3)
print(f"ThreadPool full dict:     mean={bench_threads_full['mean']:.4f}s  std={bench_threads_full['std']:.4f}s  (over {bench_threads_full['n_runs']} runs)")

# Uncomment to see the times of each run
# times_threads_full

#### Minimal dictionary

In [ ]:
def preprocess_threads_min(paths, size=(224,224), bins=32, limit=None):
    """
    Preprocess pipeline using ThreadPoolExecutor.
    Returns a minimal dictionary of features.
    """
    if limit: paths = paths[:limit]

    with ThreadPoolExecutor() as executor:
        results = list(executor.map(lambda path: preprocess_one_min(path, size, bins), paths))

    return results

bench_threads_min, times_threads_min = benchmark_function(preprocess_threads_min, paths, limit=SAMPLE_SIZE, n_runs=3)
print(f"ThreadPool minimal dict:  mean={bench_threads_min['mean']:.4f}s  std={bench_threads_min['std']:.4f}s  (over {bench_threads_min['n_runs']} runs)")

# Uncomment to see the times of each run
# times_threads_min

### 2.5 - Evaluating Speedup

All methods benchmarked above are now compared on a common table.  
Serial means (over 20 runs) are used as the reference baseline.  
Parallel and thread means are over 3 runs each; fewer runs to limit memory pressure, but enough to smooth out scheduling noise.

In [ ]:
serial_time     = bench_serial_full['mean']
serial_min_time = bench_serial_min['mean']

print(f"{'Method':<30} {'Mean (s)':>10} {'Std (s)':>10} {'Speedup vs serial':>20}")
print("-" * 74)
print(f"{'Serial (full dict)':<30} {serial_time:>10.4f} {bench_serial_full['std']:>10.4f} {'1.00x':>20}")
print(f"{'Serial (minimal dict)':<30} {serial_min_time:>10.4f} {bench_serial_min['std']:>10.4f} {'1.00x':>20}")
print(f"{'Processes (full dict)':<30} {bench_proc_full['mean']:>10.4f} {bench_proc_full['std']:>10.4f} {serial_time / bench_proc_full['mean']:>19.2f}x")

print(f"{'Processes (minimal dict)':<30} {bench_proc_min['mean']:>10.4f} {bench_proc_min['std']:>10.4f} {serial_min_time / bench_proc_min['mean']:>19.2f}x")print(f"{'Threads (minimal dict)':<30} {bench_threads_min['mean']:>10.4f} {bench_threads_min['std']:>10.4f} {serial_min_time / bench_threads_min['mean']:>19.2f}x")
print(f"{'Threads (full dict)':<30} {bench_threads_full['mean']:>10.4f} {bench_threads_full['std']:>10.4f} {serial_time / bench_threads_full['mean']:>19.2f}x")

**Expected observations** (exact numbers will vary by hardware):

- **Processes** should be faster than serial. Each worker runs on a separate core, bypassing the GIL.  
  The minimal dict variant should show a bigger speedup than the full dict, because smaller return values mean less data to pickle and send back to the main process.

- **Threads** should be *slower* than serial for this workload.  
  Image preprocessing is CPU-bound: threads compete for the GIL, context-switch constantly, and add scheduling overhead without gaining any parallel CPU time.  
  A speedup below 1.0 here is expected and informative. It confirms the GIL's impact on CPU-bound code.

### 2.6 - Batch Preprocessing using ProcessPoolExecutor

Instead of preprocessing one image per task, batches of images will be preprocessed.  
This reduces the number of tasks submitted to the pool, which means:
- Less overhead: fewer scheduling, context switching, and inter-process communication events.
- More efficient computation: operations like image decoding, resizing or normalization are often faster when applied to a batch instead of individual items.

In [ ]:
def worker(batch, size, bins):
    return [preprocess_one_min(p, size=size, bins=bins) for p in batch]

def preprocess_parallel_batches(paths, size=(224,224), bins=32, batch_size=16, max_workers=None, limit=None):
    if limit:
        paths = paths[:limit]

    batches = [paths[i:i+batch_size] for i in range(0, len(paths), batch_size)]

    time_start = time.perf_counter()
    first_future_submit = None
    first_result_time = None

    results = []
    with ProcessPoolExecutor(max_workers=max_workers) as executor:
        futures = []
        for batch in batches:
            if first_future_submit is None:
                first_future_submit = time.perf_counter()
            futures.append(executor.submit(worker, batch, size, bins))
        for i, f in enumerate(as_completed(futures)):
            result = f.result()
            if first_result_time is None:
                first_result_time = time.perf_counter()
            results.extend(result)

    time_end = time.perf_counter()

    return {
        "n_images": len(results),
        "total_time": time_end - time_start,
        "submit_overhead": (first_future_submit - time_start) if first_future_submit else None,
        "time_until_first_result": (first_result_time - time_start) if first_result_time else None,
        "batch size": batch_size,
        "max_workers": max_workers
    }


In [ ]:
stats_batch = preprocess_parallel_batches(paths, limit=SAMPLE_SIZE, batch_size=8)
stats_batch

#### Explore the effect of batchsize on the preprocessing with ProcessPoolExecutor

In [ ]:
batch_sizes = [4, 8, 16, 24, 32, 48]
sweep = []
for b in batch_sizes:
    s = preprocess_parallel_batches(paths, limit=SAMPLE_SIZE, batch_size=b)
    sweep.append(s)
    print(f"batch={b:2d} total={s['total_time']:.3f}s first_result={s['time_until_first_result']:.3f}s")

# Find best
best = min(sweep, key=lambda d: d['total_time'])
print("\nBest batch config:", best)

In [ ]:
total_times = [d['total_time'] for d in sweep]

plt.plot(batch_sizes, total_times, marker='o')
plt.title("Total Time vs Batch Size")
plt.ylabel("Total Time (s)")
plt.xlabel("Batch Size")
plt.ylim(0, max(total_times) * 1.1)
plt.xticks(batch_sizes);


#### Processes with Best Batchsize vs Serial and Processes without batch

In [ ]:
serial_min_time       = bench_serial_min['mean']
parrallel_min_time    = bench_proc_min['mean']

print(f"Serial minimal dict ({SAMPLE_SIZE} images): {serial_min_time:.4f}s")
print(f"Processes no-batch  ({SAMPLE_SIZE} images): {parrallel_min_time:.4f}s")

print(f"\nSpeedup — batched processes vs serial:      {serial_min_time / best['total_time']:.2f}x")
print(f"Speedup — batched processes vs no-batch:    {parrallel_min_time / best['total_time']:.2f}x")

Using a combination of batches and processes, with a batch size of 16, increased the preprocessing speed by 1.90 compared to serial preprocessing, and by 1.18 for processes without batches.  

Batching increases preprocessing speed mainly because there is less overhead per unit of useful work.  

Details:  
- Pure processes (no batching)
  - Each image is handed off to a worker process individually.
  - This means a lot of overhead in:
    - pickling/unpickling data to send it to the worker
    - process synchronization (waiting for tasks/results)
    - OS scheduling overhead when workers finish small tasks quickly
- Processes + batching
  - Each worker gets a chunk of images to process in one go.
  - That means fewer messages sent between parent and child processes.
  - Workers spend more time doing actual work (preprocessing) vs. waiting for the next task.
  - Better CPU cache locality: operating on a batch of similar data often plays nicely with memory/cache.

## 3 - Feature extraction

### 3.1 - Using a pretrained CNN (MobileNetV2)

[MobileNetV2](https://arxiv.org/abs/1801.04381) is a lightweight CNN architecture pretrained on ImageNet.  
We use it as a **feature extractor**: the final classification head is removed (`include_top=False`) and replaced with global average pooling, producing a **1280-dimensional feature vector** per image that encodes semantic content (shapes, textures, object parts).

Using a pretrained model means we skip training entirely and still get meaningful, high-quality representations, useful for downstream tasks like clustering or classification.

MobileNetV2 applies its own input normalisation (pixel values scaled to `[-1, 1]`), so the custom preprocessing pipeline from Section 2 is not used here.

#### Extract features one image at a time

In [ ]:
def extract_features_single(path, size=(224,224)):
    """
    Loads an image, resizes, and extracts features using MobileNetV2.
    Returns a 1280-dim feature vector.
    """

    # Load and preprocess image
    with Image.open(path) as image:
        image = image.convert("RGB").resize(size, Image.BILINEAR)
        image_resized = np.array(image, dtype=np.float32)
    image_resized = tf.keras.applications.mobilenet_v2.preprocess_input(image_resized)
    image_resized = np.expand_dims(image_resized, axis=0)  # shape (1, H, W, 3)

    features = extract_features_single.model(image_resized)

    return features.numpy().squeeze()

#### Extract features from a batch of images

#### 3.1.2 - Batch extraction (parallel, one model per process)

TensorFlow creates CUDA contexts and internal runtime state that cannot be pickled or safely shared across process boundaries.  
The solution: **each worker process loads its own copy of the model** at the start of the batch.  
This adds a one-time model-loading overhead per process, but avoids all cross-process TensorFlow issues.

In [ ]:
def batch_extract_features(paths, size=(224,224)):
    """
    Takes a batch of image paths.
    Loads MobileNetV2 once per process to extract the features of each image.
    Returns a list of feature vectors.
    """

    # Load model
    model = tf.keras.applications.MobileNetV2(
        include_top=False, pooling='avg', input_shape=(size[0], size[1], 3)
    )

    out = []

    for path in paths:
        with Image.open(path) as img:
            img = img.convert("RGB").resize(size, Image.BILINEAR)
            image_resized = np.array(img, dtype=np.float32)
        image_resized = tf.keras.applications.mobilenet_v2.preprocess_input(image_resized)
        image_resized = np.expand_dims(image_resized, axis=0)
        features = model(image_resized)
        out.append(features.numpy().squeeze())

    return out

#### Testing the batch_extract_features function inside parallel_map_batched

Sample size:  100  
Batch size:   8

In [ ]:
@time_it
def extract_features():
    features_100 = parallel_map_batched(
        paths,
        batch_extract_features,
        is_batched=True,
        batch_size=8,
        max_workers=None,
        func_kwargs={'size': (224,224)},
        limit=100,
        display_progress=True
    )
    return features_100
features_100 = extract_features()
print(f"Extracted features for {len(features_100)} images. Feature shape: {features_100[0].shape}")

Sample size:  100  
Batch size:   16

In [ ]:
@time_it
def extract_features():
    features_100 = parallel_map_batched(
        paths,
        batch_extract_features,
        is_batched=True,
        batch_size=16,
        max_workers=None,
        func_kwargs={'size': (224,224)},
        limit=100,
        display_progress=True
    )
    return features_100
features_100 = extract_features()
print(f"Extracted features for {len(features_100)} images. Feature shape: {features_100[0].shape}")

Sample size:  1000  
Batch size:   32

In [ ]:
@time_it
def extract_features():
    features_big = parallel_map_batched(
        paths,
        batch_extract_features,
        is_batched=True,
        batch_size=32,
        max_workers=None,
        func_kwargs={'size': (224,224)},
        limit=1000,
        display_progress=True
    )
    return features_big
features_big = extract_features()
print(f"Extracted features for {len(features_big)} images. Feature shape: {features_big[0].shape}")

In [ ]:
features_arr = np.stack(features_big)
print("Feature array shape:", features_arr.shape)

## 4 - Features visualization

Each image is now represented by a 1280-dimensional vector, which is impossible to visualise directly.  
[PCA (Principal Component Analysis)](https://en.wikipedia.org/wiki/Principal_component_analysis) finds the two directions of maximum variance in the feature space and projects all vectors onto them.  
This gives a 2D view that preserves as much structure as possible, letting us visually check whether images from the same class cluster together.

In [ ]:
from sklearn.decomposition import PCA

# Reduce number of dimensions with PCA to 2D for visualization
pca = PCA(n_components=2)
features_2d = pca.fit_transform(features_arr)

In [ ]:
print("Reduced feature shape:", features_2d.shape)

In [ ]:
import matplotlib.pyplot as plt

plt.figure(figsize=(8,6))
plt.scatter(features_2d[:,0], features_2d[:,1], s=10, alpha=0.7)
plt.title("CNN Features (PCA 2D projection)")
plt.xlabel("PC1")
plt.ylabel("PC2")
plt.show()

In [ ]:
import pandas as pd

labels = [str(p).split('/')[-2] for p in paths[:len(features_big)]]
unique_labels = list(pd.Series(labels).unique())
label_to_idx = {lbl: i for i, lbl in enumerate(unique_labels)}
colors = [label_to_idx[lbl] for lbl in labels]

print(label_to_idx)

plt.figure(figsize=(8,6))
plt.scatter(features_2d[:,0], features_2d[:,1], c=colors, cmap='tab10', s=10, alpha=0.7)
plt.title("CNN Features by Class (PCA 2D projection)")
plt.xlabel("PC1")
plt.ylabel("PC2")
plt.colorbar()
plt.show()

#### Clustering with Kmeans

In [ ]:
from sklearn.cluster import KMeans

n_clusters = len(unique_labels)
kmeans = KMeans(n_clusters=n_clusters)
cluster_labels = kmeans.fit_predict(features_arr)

plt.figure(figsize=(8,6))
plt.scatter(features_2d[:,0], features_2d[:,1], c=cluster_labels, cmap='tab10', s=10, alpha=0.7)
plt.title(f"KMeans Clusters (k={n_clusters}) in CNN Feature Space")
plt.xlabel("PC1")
plt.ylabel("PC2")
plt.colorbar(label="Cluster")
plt.show()

## 5 - End‑to‑End Pipeline: CPU Loader Workers => Single Model Inference (Queue)

In [ ]:
import multiprocessing as mp
from queue import Empty

SENTINEL = None  # marks end of stream

def loader_worker(worker_id, in_paths, out_queue, size=(224,224)):
    for p in in_paths:
        try:
            with Image.open(p) as img:
                img = img.convert("RGB").resize(size, Image.BILINEAR)
                arr = np.array(img, dtype=np.uint8)
        except Exception as e:
            continue
        out_queue.put((str(p), arr))
    out_queue.put(SENTINEL)

def run_pipeline(paths, n_loader_procs=4, size=(224,224), batch_size=8, limit=1000):
    paths = paths[:limit]
    t0 = time.perf_counter()
    # Split paths roughly evenly
    chunks = [paths[i::n_loader_procs] for i in range(n_loader_procs)]
    q = mp.Queue(maxsize=batch_size * 4)
    procs = [
        mp.Process(target=loader_worker, args=(i, chunk, q, size))
        for i, chunk in enumerate(chunks)
    ]
    for p in procs:
        p.start()

    # Single model (avoid multiple heavy loads)
    model = tf.keras.applications.MobileNetV2(
        include_top=False, pooling='avg', input_shape=(size[0], size[1], 3)
    )

    finished = 0
    features = []
    path_order = []

    def preprocess_batch(batch_imgs):
        batch_arr = np.stack(batch_imgs).astype(np.float32)
        batch_arr = tf.keras.applications.mobilenet_v2.preprocess_input(batch_arr)
        return model(batch_arr).numpy()

    current_imgs = []
    current_paths = []

    while finished < n_loader_procs:
        try:
            item = q.get(timeout=1.0)
        except Empty:
            continue
        if item is SENTINEL:
            finished += 1
            continue
        pth, img_arr = item
        current_imgs.append(img_arr)
        current_paths.append(pth)
        if len(current_imgs) == batch_size:
            feats = preprocess_batch(current_imgs)
            features.append(feats)
            path_order.extend(current_paths)
            current_imgs.clear()
            current_paths.clear()

    # Flush remainder
    if current_imgs:
        feats = preprocess_batch(current_imgs)
        features.append(feats)
        path_order.extend(current_paths)

    for p in procs: p.join()

    feats_all = np.vstack(features)
    t1 = time.perf_counter()
    return {
        "features": feats_all,
        "paths": path_order,
        "time": t1 - t0,
        "n_images": len(path_order),
        "throughput_img_per_s": len(path_order)/(t1 - t0)
    }

pipeline_stats = run_pipeline(paths, n_loader_procs=4, size=(224,224), batch_size=8, limit=SAMPLE_SIZE)
pipeline_stats

In [ ]:
# Compare with previous parallel_map_batched result (features_big already computed)
prev_time = None  # set if you stored timing for features_big earlier
print("Pipeline images:", pipeline_stats["n_images"])
print("Pipeline time (s):", f"{pipeline_stats['time']:.2f}")
print("Pipeline throughput (img/s):", f"{pipeline_stats['throughput_img_per_s']:.1f}")
if prev_time:
    print("Speedup vs previous multi-process model-per-process:", f"{prev_time / pipeline_stats['time']:.2f}x")